### PR TITLE
PEDS-1168 Added notif schema

### PR DESCRIPTION
**Pull request by okmbaye**

### CODE REVIEW

File: migrations/versions/57ef835d5488_add_and_track_user_notifications.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+revision = '57ef835d5488'
+down_revision = '91b6a04820cd'
+branch_labels = None
+depends_on = None
+
+
+def upgrade() -> None:
+    op.create_table('notification', 
+                    sa.Column('notification_id', sa.Integer(), sa.ForeignKey("notification_log.id"), nullable = False), 
+                              sa.Column('user_id', sa.Integer(), sa.ForeignKey("associated_user.user_id"), nullable = False),
```
Comment (grugna) [inline migrations/versions/57ef835d5488_add_and_track_user_notifications.py:22]: this is just integer, no foreign key

**Comment by grugna** [comment]
@okmbaye you need to import the models ( https://github.com/chicagopcdc/amanuensis/blob/483b4e78aff75d705ad27e8eee82782174d731c5/amanuensis/models.py#L25 ) otherwise they will not be found. 

Did you run this?

**Comment by grugna** [comment]
@okmbaye make the injectnotification and retrievenotification one single blueprint instead of 2. And call it notification

File: amanuensis/resources/userdatamodel/userdatamodel_notification.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+
+# Get all notifications associated with a user
+def get_notifications_by_user(current_session, logged_user_id):
+    notifications = current_session.query(Notification).filter_by(user_id = logged_user_id)
+    return notifications
+
+
+# Get all notifications associated with a log
+def get_log_assoiciated(current_session, log_id):
```
Comment (grugna) [inline amanuensis/resources/userdatamodel/userdatamodel_notification.py:45]: typo *associated

File: amanuensis/resources/notification/__init__.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+# Get the notifications that are not seen
+def get_unseen_notifications(users):
+    unseen = []
+    with flask.current_app.db.session as session:
+        update_notifications()
+        for user in users:
+            user_unseen = check_unseen_notifications(session, user)
+            unseen.extend(user_unseen)
+
+        return unseen
```
Comment (grugna) [inline amanuensis/resources/notification/__init__.py:106]: this should be:

1. call to userdatamodel.get_notifications_by_user (this is the value to be returned)
2. Set all the rows retrieved to seen = True
3. return the list of values at point 1

File: amanuensis/resources/userdatamodel/userdatamodel_notification.py (hunk 2 above)
Comment (grugna) [inline amanuensis/resources/userdatamodel/userdatamodel_notification.py:45]: both this functions should have a seen parameter defaulted to null. If the parameter is not null then you should append to the query WHERE seen = ...

File: amanuensis/blueprints/retrieve_notifications.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+@blueprint.route("/", methods=["GET"])
+def get_all_users():
+    user_notifications = notification.get_notifications()
+    users = []
+    for notif in user_notifications:
+        if notif.user_id not in users:
+            users.append(notif.user_id)
+    
+    return users
```
Comment (grugna) [inline amanuensis/blueprints/retrieve_notifications.py:33]: @okmbaye what is the meaning of this? retrieving all the users related to any notification?
If this is correct, you can remove this endpoint.

File: amanuensis/blueprints/inject_notifications.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+@blueprint.route("/", methods=["GET"])
+def get_all_users():
+    user_notifications = notification.get_notifications()
+    users = []
+    for notif in user_notifications:
+        if notif.user_id not in users:
+            users.append(notif.user_id)
+    
+    return users
```
Comment (grugna) [inline amanuensis/blueprints/inject_notifications.py:25]: If this is just to retrieve all the users that have at least 1 notification I don't think I can see the meaning. What is this for?

File: amanuensis/blueprints/inject_notifications.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+
+@blueprint.route("/", methods=["POST"])
```
Comment (grugna) [inline amanuensis/blueprints/inject_notifications.py:28]: add admin authZ

File: amanuensis/blueprints/inject_notifications.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+def add_notifications(messages, user_ids):
```
Comment (grugna) [inline amanuensis/blueprints/inject_notifications.py:29]: what is the meaning of user_ids here? When a new notification is created should be broadcasted to anyone for how this has been designed. remove user_ids

File: amanuensis/blueprints/inject_notifications.py (hunk 6 above)
Comment (grugna) [reply amanuensis/blueprints/inject_notifications.py:28]: this method should be inside the admin blueprints

File: amanuensis/blueprints/retrieve_notifications.py [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+blueprint = flask.Blueprint("retriveNotification", __name__)
+
+logger = get_logger(__name__)
+
+@blueprint.route("/", methods=["GET"])
+def get_notifications():
+    user_notifications = notification.get_notifications()
+    notification_schema = NotificationSchema(many=True)
+    
+    return flask.jsonify(notification_schema.dump(user_notifications))
```
Comment (grugna) [inline amanuensis/blueprints/retrieve_notifications.py:22]: this should also be only for admin and be in the admin blueprints

File: amanuensis/blueprints/retrieve_notifications.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    
+    return users
+
+
+@blueprint.route("/", methods=["GET"])
+def get_unseen_notifications(users):
+    get_unseen = notification.get_unseen_notifications(users)
+    unseen_schema = NotificationSchema(many=True)
+    
+    return flask.jsonify(unseen_schema.dump(get_unseen))
```
Comment (grugna) [inline amanuensis/blueprints/retrieve_notifications.py:41]: this need the authN and retrieve the user_id of the logged in user

File: amanuensis/resources/notification/__init__.py [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    notificationlog_schema = NotificationLogSchema(many = True)
+    new_logs = []
+    with flask.current_app.db.session as session:
+        for message in messages:
+            new_notifLog, new_notif_for_users = create_new_notification(session, user_ids, message)
+            notificationlog_schema.dump(new_notifLog)
+            notification_schema.dump(new_notif_for_users)
+            new_logs.append(new_notifLog)
+
+        return new_logs
```
Comment (grugna) [inline amanuensis/resources/notification/__init__.py:69]: the new notification should be added only in the notification log table

**Comment by grugna** [comment]
Add some test please.

**Comment by paulmurdoch19** [comment]
@okmbaye you will need to register the new notification blueprint here https://github.com/chicagopcdc/amanuensis/blob/6e80733a0c7f3516f2e1837105e74eff3c3b4a97/amanuensis/__init__.py#L79

**Comment by paulmurdoch19** [comment]
@grugna Yes I can do that

**Comment by paulmurdoch19** [comment]
@grugna I updated this to follow the updated format for the service but the logic should be the same. This https://github.com/chicagopcdc/userPortalDataModel/pull/52 will need to be merged in before this one.

**Comment by grugna** [comment]
@paulmurdoch19 I'll review it now. the new userportaldatamodel version is 1.8.0

**Comment by grugna** [comment]
@paulmurdoch19 also have you already merged in the latest from pcdc_dev branch?

**Comment by grugna** [comment]
@paulmurdoch19 this looks good, there is only one thing I would like to change before merging. I realized as it is designed now if we create for instance 20 notifications during 1 year, then a new user on year 2 will receive all 20 of them and if the idea is using them to notify people of deployment it will be useless information because it will display notifications about last year deployments.

So we should add an expiration_date col in the notification_log table. On the create notification log endpoint we should also pass an expiration_date. And in the non admin GET /notifications endpoint we should filter out the expired notifications automatically, while on the admin endpoint it should be a parameter passed to the function, so we could retrieve also the expired ones.

**Comment by paulmurdoch19** [comment]
@grugna Thats a good idea I will add that and yes that branch is up to date with pcdc dev

File: amanuensis/blueprints/notification.py [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+logger = get_logger(__name__)
+
+blueprint = flask.Blueprint("notifications", __name__)
+
+
+
+@blueprint.route("/", methods=["GET"])
+def retrieve_notifications():
+    new = flask.request.args.get("new", type=bool, default=False)
```
Comment (grugna) [inline amanuensis/blueprints/notification.py:17]: this is unused so can be removed

**Comment by paulmurdoch19** [comment]
this needs to be added before merge https://github.com/chicagopcdc/userPortalDataModel/pull/59

**Comment by grugna** [comment]
new userportaldatamodel version is 1.8.1

**Pull request merged**